sklearn/cluster/optics_.py

```diff
@@ -527,3 +568,382 @@ def cluster_optics_dbscan(reachability, core_distances, ordering, eps=0.5):
+        constitutes a cluster boundary. For example, an upwards point in the
+        reachability plot is defined by the ratio from one point to its
+        successor being at most 1-xi.
+
+    min_samples : integer
+       The same as the min_samples given to OPTICS. Up and down steep
+       regions can't have more then `min_samples` consecutive non-steep
+       points.
+
+    min_cluster_size : int > 1 or float between 0 and 1
+        Minimum number of samples in an OPTICS cluster, expressed as an
+        absolute number or a fraction of the number of samples (rounded
+        to be at least 2).
+
+    predecessor_correction : bool
+        Correct clusters based on the calculated predecessors.
+
+    Returns
+    -------
+    clusters : array, shape (n_clusters, 2)
+        The list of clusters in the form of [start, end] in each row, with all
+        indices inclusive. The clusters are ordered in a way that larger
+        clusters encompassing smaller clusters come after those smaller
+        clusters.
+    """
+
+    # all indices are inclusive (specially at the end)
+    n_samples = len(reachability_plot)
+    # add an inf to the end of reachability plot
+    # but process the data only until the last actual point
+    # this is fine since the last point is considered upward anyway
+    reachability_plot = np.hstack((reachability_plot, np.inf))
+
+    if min_cluster_size <= 1:
+        min_cluster_size = max(2, min_cluster_size * n_samples)
+
+    xi_complement = 1 - xi
+    sdas = []  # steep down areas, introduced in section 4.3.2 of the paper
+    clusters = []
+    index = 0
+    mib = 0.  # maximum in between
+
+    ratio = reachability_plot[:-1] / reachability_plot[1:]
+    steep_upward = ratio <= xi_complement
+    steep_downward = ratio >= 1 / xi_complement
+    downward = ratio > 1
+    upward = ratio < 1
+
+    # the following loop is is almost exactly as Figure 19 of the paper.
+    # it jumps over the areas which are not either steep down or up areas
+    for steep_index in iter(np.flatnonzero(steep_upward | steep_downward)):
+        # just continue if steep_index has been a part of a discovered xward
+        # area.
+        if steep_index < index:
+            continue
+
+        mib = max(mib, np.max(reachability_plot[index:steep_index + 1]))
+        index = steep_index
+
+        # check if a steep downward area starts
+        if steep_downward[index]:
+            sdas = _update_filter_sdas(sdas, mib, xi_complement)
+            D_start = index
+            index, end = _extend_downward_region(steep_downward, upward,
+                                                 D_start, min_samples)
+            D = _Area(start=D_start, end=end,
+                      maximum=reachability_plot[D_start], mib=0.)
+            sdas.append(D)
+            mib = reachability_plot[end + 1]
+
+        elif steep_upward[index]:
+            sdas = _update_filter_sdas(sdas, mib, xi_complement)
```

Comment:
so you don't consider the points in the current steep up region (except for the first point)? (i.e., you only check sc1* for ``sD < x <= sU``)?

Reply:
according to Figure 19 of the paper, sdas are updated before changing the `index, and that's why I've implemented like this.